### PR TITLE
[ADD] Upgrade documentation for developers

### PR DESCRIPTION
**Pull request by hepi-odoo**
More and more demand came from partners to have information about the upgrade process.
In this page, we intend to introduce the process we've been following at PS-tech for several year.
In the future there will be multiple pages around this one giving explanation on upgrade related topics.

### CODE REVIEW

File: content/developer/howtos/upgrade_custom_db.rst [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+
+Upgrading to a new version of Odoo can be challenging, 
+especially if the database you work on contains custom modules.  
+This page intent is to explain the technical process of upgrading a database with customized modules.
+For numerous years, Odoo has followed this process, reaping the rewards of continual improvements along the way.
+
+For a standard overview, please refer to the
+:doc:`upgrade documentation </administration/upgrade>`.
+
+We consider a custom module, any module that extends the standard code of Odoo and that was not built with studio.
```
Comment (augustoPerez) [inline content/developer/howtos/upgrade_custom_db.rst:19]: Could it be better to say?
```suggestion
We consider a custom module, any module that extends the standard code of Odoo and that was not built with the Studio app. 
```

File: content/developer/howtos/upgrade_custom_db.rst [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+Stop the developments
+=============================
+
+Upgrade can be a tough process, starting it requires commitment and will block a significant amount of your development resources.
+If you decide to keep going for developments during that process,
+you will inevitably have to block resources for a longer time to work on the upgrade.
+Those features will need to be re-upgraded and tested, everytime you change them.
+
+This is why we recommend a complete freeze of the codebase when starting the upgrade process.
+Of course, you can still work on maintenance but developing new features is not productive.
```
Comment (augustoPerez) [inline content/developer/howtos/upgrade_custom_db.rst:62]: Maybe `bugfix` instead of maintenance?
That how we call it but IMO, "bugfix" is more understandable and means a smaller scope.
```suggestion
Of course, you can still work on bigfix but developing new features is not productive.
```

File: content/developer/howtos/upgrade_custom_db.rst [hunk 3]
```diff
@@ -0,0 +1,305 @@
...
+.. database and their code clean.
+.. It is ready to make the custom modules installable on the upgraded database.
+
+.. _upgrade/empty_database/modules_installable:
+
+Make custom modules installable
+-------------------------------
+
+The first step is to make the custom modules installable in the new Odoo version.
+This means, in a first instance, making sure there is no traceback or warnings when installing them.
```
Comment (hepi-odoo) [inline content/developer/howtos/upgrade_custom_db.rst:132]: I'm not sure about this sentence (all though your english is better than mine)

File: content/developer/howtos/upgrade_custom_db.rst [hunk 4]
```diff
@@ -0,0 +1,305 @@
...
+no errors during their execution.
+Because of this, we encourage to test thoroughly all the customization to make sure everything
+is working as expected.
+
+This process will help detect further issues that are not identified during the module installation
+and can only be detected in runtime.
+For example, deprecated calls to standard python or OWL functions, non existing references to
+standard fields, etc.
+
+.. TODO Expand in the tests and their explanation
```
Comment (hepi-odoo) [inline content/developer/howtos/upgrade_custom_db.rst:158]: IMO, we don't really need to. Any developer knows what testing is, I think that specifying what they should look at is enough.

File: content/developer/howtos/upgrade_custom_db.rst [hunk 5]
```diff
@@ -0,0 +1,305 @@
...
+.. _upgrade/empty_database/clean_code:
+
+Clean the code
+--------------
+
+At this stage of the upgrade process, we also suggest to clean the code as much as possible.
+This includes: 
+
+  - Remove redundant and unnecessary code.
+  - Remove features that are now part of Odoo standard.
```
Comment (hepi-odoo) [inline content/developer/howtos/upgrade_custom_db.rst:183]: This is already explained in section "stop the development

File: content/developer/howtos/upgrade_custom_db.rst [hunk 6]
```diff
@@ -0,0 +1,314 @@
...
+empty database in the targeted version of your upgrade.
+
+This ensures that the customization is compatible with the new version of Odoo, allows to analyse
+how does it behave and interact with the new features, and guarantees that they will not cause
+any issue when upgrading the database.
+
+Making the custom modules work in an empty database also helps avoiding changes and wrong
+configurations that might be present on the production database (like studio customization,
+customized website pages, mail templates or translations). They are not intrinsically related to the
+custom modules and that can raise unwanted issues in this stage of the upgraded process.
```
Comment (hepi-odoo) [inline content/developer/howtos/upgrade_custom_db.rst:113]: We could make this more convicing

File: content/developer/howtos/upgrade_custom_db.rst (hunk 4 above)
Comment (augustoPerez) [reply content/developer/howtos/upgrade_custom_db.rst:160]: I'm not sure why I added this :thinking: 
That means we can remove it

File: content/developer/howtos/upgrade_custom_db.rst (hunk 5 above)
Comment (augustoPerez) [reply content/developer/howtos/upgrade_custom_db.rst:192]: Should we include a reference to that part of the doc or remove this point?

Comment (hepi-odoo) [reply content/developer/howtos/upgrade_custom_db.rst:192]: Absolutely!

File: content/developer/howtos/upgrade_custom_db.rst [hunk 7]
```diff
@@ -0,0 +1,287 @@
...
+errors during their execution. Because of this, we encourage to test thoroughly all the
+customization to make sure everything is working as expected.
+
+This process will help detect further issues that are not identified during the module installation
+and can only be detected in runtime. For example, deprecated calls to standard python or OWL
+functions, non existing references to standard fields, etc.
+
+We recommend to test all the customization, specially the following elements:
+
+- Views
```
[the remark this answers is not in the document]
Comment (augustoPerez) [reply content/developer/howtos/upgrade_custom_db.rst:149]: Maybe not fail at runtime, but maybe the elements added are not displayed anymore in the place they are intended compared to the previous version, so I recommend to test it too

File: content/developer/reference/backend/migration_scripts.rst [hunk 8]
```diff
@@ -0,0 +1,134 @@
+.. _upgrade/migration-scripts:
```
[the remark this answers is not in the document]
Comment (augustoPerez) [reply content/developer/reference/backend/migration_scripts.rst:1]: Maybe we can check with the marketing team, but it probably should be removed :confused:

File: content/developer/reference/cli.rst [hunk 9]
```diff
@@ -136,13 +136,9 @@ Running the server
```
Comment (augustoPerez) [inline content/developer/reference/cli.rst:136]: Hi @nmarotte 
Why was this file changed for the upgrade documentation?

Comment (nmarotte) [reply content/developer/reference/cli.rst:136]: I think some other changes got absorbed into my commit by accident. I modified this file to change the text related to the odoo shell, starting at "This shell is started with the `env` variable", as well as the first paragraph of the section.

I don't remember the rest of the changes for that file so I think they are probably from another commit @augustoPerez

**Comment by AntoineVDV** [comment]
Hi there, I see that the PR is still a work in progress (TODOs in the content, "Temporary PR" in the description...) but it was made ready for review; was it intended?

File: content/developer/reference/cli.rst (hunk 9 above)
Comment (augustoPerez) [reply content/developer/reference/cli.rst]: @nmarotte I adapted to keep just your changes, could you check if what I did is correct and nothing got lost?
Thanks!

Comment (nmarotte) [reply content/developer/reference/cli.rst]: @augustoPerez perfect, looks good to me :D

**Comment by hepi-odoo** [comment]
Hello @AntoineVDV , we might have put it a few hours too early for review but it's ready now

File: content/developer/howtos/upgrade_custom_db.rst [hunk 10]
```diff
@@ -0,0 +1,281 @@
...
+
+.. _upgrade_custom/stop_developments:
+
+Stop the developments
+=====================
+
+Starting an upgrade requires commitment and development resources need to be allocated to it. If
+developments keep being made at the same time, those features will need to be re-upgraded and tested
+everytime you change them. This is why we recommend a complete freeze of the codebase when starting
+the upgrade process. Needless to mention that bug fixing is exempt from this recommendation.
```
Comment (nmarotte) [inline content/developer/howtos/upgrade_custom_db.rst:58]: If you are using Needless, I would personally write it this way:  `Needless to say, bug fixing is exempt from this recommendation.`, feels more natural to me than `Needless to mention that`

File: content/developer/howtos/upgrade_custom_db.rst [hunk 11]
```diff
@@ -0,0 +1,281 @@
...
+
+Once you have stopped development, it is a good practice to assess the developments made and compare
+them with the features introduced between you current version and the version you are targeting.
```
Comment (nmarotte) [inline content/developer/howtos/upgrade_custom_db.rst:61]: between you**r** (r is missing)

File: content/developer/howtos/upgrade_custom_db.rst [hunk 12]
```diff
@@ -0,0 +1,281 @@
...
+Challenge the developments as much as possible, find functionnal workarounds.
+Removing redundancy between your developments and the standard version of Odoo will lead to an eased
+upgrade process and reduce the technical debt.
+
+.. note::
+   You can find information on the changes between versions in the `release notes
+   <https:/odoo.com/page/release-notes>`_. Those may seem light at first but they will help point
```
Comment (nmarotte) [inline content/developer/howtos/upgrade_custom_db.rst:68]: The sentence `Those may seem light at first but ...` stood out when reading this, IMO it clashes with the style of the other sentences and the global tone of the page. I don't think we should introduce any biais towards the release notes and let the reader decide if they feel like they are light or not IMO

File: content/developer/howtos/upgrade/migration_scripts.rst [hunk 13]
```diff
@@ -0,0 +1,108 @@
...
+
+#. What they are (python scripts that receives the psql cursor and the version)
+#. What they are for : applying a modification on the data of your database when upgrading a module
+#. In what circumstances you should use them : when changing the source code of your module between
+   2 versions of Odoo
+#. How to write them : examples, upgrade-util package
+#. The different phases and what is their impact
+#. Where to put them : in <module_name>/migrations
+#. How to test them Odoo SH : branch in upgrade mode, push a commit, On-premise: receive your dump,
+   run it and upgrade all modules via starting odoo-bin
```
[the remark this answers is not in the document]
Comment (augustoPerez) [reply content/developer/howtos/upgrade/migration_scripts.rst:17]: I'm about to remove the file from this PR and we'll add it to the next one (maybe togheter with a `util` one, or all-in-one file)

File: content/developer/howtos/upgrade_custom_db.rst [hunk 14]
```diff
@@ -0,0 +1,281 @@
...
+
+
+.. _upgrade_custom/request_upgrade:
+
+Request an upgraded database
+============================
+
+Once the developments have stopped for the custom modules and the implemented features have been
+challenged to remove redundancy and unnecessary code, the next step is to request an upgraded test
+database. To request the upgraded test database, follow the steps mentioned in
```
Comment (nmarotte) [inline content/developer/howtos/upgrade_custom_db.rst:79]: `... the next step is to request an upgraded test database. To request the upgraded test database ...` maybe we could start the next sentence with `To do so, ...` to avoid repetition ? :D

File: content/developer/howtos/upgrade_custom_db.rst [hunk 15]
```diff
@@ -0,0 +1,281 @@
...
+- Methods renamed or removed.
+- ...
+
+.. _upgrade_custom/empty_database/test_fixes:
+
+Test and fixes
+--------------
+
+Once there are no more tracebacks when installing the modules, the next step is to test them.
+Even if the custom modules are installable on an empty database, this does not warranty there are no
```
Comment (nmarotte) [inline content/developer/howtos/upgrade_custom_db.rst:139]: I think this should be `this does not warrant ...` as warranty is the common word, while warrant is the verb :D

**Review by nmarotte — COMMENTED** [review]
Hey :D Quickly glanced over and noticed a few things that you might want to check :)

File: content/developer/howtos/upgrade_custom_db.rst [hunk 16]
```diff
@@ -0,0 +1,262 @@
...
+
+.. _upgrade_custom/upgrade_custom_db:
+
+===============================
+Upgrading a customized database
+===============================
+
+Upgrading to a new version of Odoo can be challenging, especially if the database you work on
+contains custom modules. This page intent is to explain the technical process of upgrading a
+database with customized modules. For a standard overview, please refer to the
```
Comment (AntoineVDV) [inline content/developer/howtos/upgrade_custom_db.rst:11]: What is a "standard overview" in this context?

Maybe: "For the functional process/perspective"?

File: content/developer/howtos/upgrade_custom_db.rst [hunk 17]
```diff
@@ -0,0 +1,262 @@
...
+:doc:`Upgrade documentation </administration/upgrade>`.
+
+We consider a custom module, any module that extends the standard code of Odoo and that was not
+built with the Studio app. 
+Before upgrading those modules, or before asking for them to be upgraded, have a look at the
```
Comment (AntoineVDV) [inline content/developer/howtos/upgrade_custom_db.rst:16]:
```suggestion
Before upgrading such a module, or before asking its upgrade, have a look at the
```

File: content/developer/howtos/upgrade_custom_db.rst [hunk 18]
```diff
@@ -0,0 +1,262 @@
...
+While working on what we refer to as the **custom upgrade** of your database, keep in mind the goals
+of an upgrade:
+
+#. Stay supported
+#. Get the latest features
+#. Enjoy the performance improvement
+#. Reduce the technical debt
+#. Benefit from security improvements
+
+With every new version of Odoo, changes are introduced. These changes can have an impact modules on
```
Comment (AntoineVDV) [inline content/developer/howtos/upgrade_custom_db.rst:28]:
```suggestion
With every new version of Odoo, changes are introduced. These changes can impact modules on
```

File: content/developer/howtos/upgrade_custom_db.rst [hunk 19]
```diff
@@ -0,0 +1,262 @@
...
+which customization have been developed. This is the reason why upgrading a database that contains
+custom modules requires additional steps in order to upgrade their source code.
```
Comment (AntoineVDV) [inline content/developer/howtos/upgrade_custom_db.rst:30]:
```suggestion
custom modules requires additional steps in order to upgrade the source code.
```
```suggestion
custom modules requires additional steps in order to upgrade its source code.
```

File: content/developer/howtos/upgrade_custom_db.rst [hunk 20]
```diff
@@ -0,0 +1,262 @@
...
+
+These are the steps to follow to upgrade customized databases:
+
+#. :ref:`Stop the devolopments and challenge them <upgrade_custom/stop_developments>`.
+#. :ref:`upgrade_custom/request_upgrade`.
+#. :ref:`Make your module installable on an empty database <upgrade_custom/empty_database>`.
+#. :ref:`Make your module installable on the upgraded database <upgrade_custom/upgraded_database>`.
+#. :ref:`Test extensively and do a rehearsal <upgrade_custom/testing_rehearsal>`.
+#. :ref:`Upgrade in production <upgrade_custom/production>`.
```
Comment (AntoineVDV) [inline content/developer/howtos/upgrade_custom_db.rst:39]: If these are really steps you could also prepend numbers to the header.

File: content/developer/howtos/upgrade_custom_db.rst [hunk 21]
```diff
@@ -0,0 +1,262 @@
...
+
+
+.. _upgrade_custom/stop_developments:
+
+Stop the developments
+=====================
+
+Starting an upgrade requires commitment and development resources need to be allocated to it. If
```
Comment (AntoineVDV) [inline content/developer/howtos/upgrade_custom_db.rst:47]:
```suggestion
Starting an upgrade requires commitment and development resources. If
```

File: content/developer/howtos/upgrade_custom_db.rst [hunk 22]
```diff
@@ -0,0 +1,262 @@
...
+developments keep being made at the same time, those features will need to be re-upgraded and tested
+everytime you change them. This is why we recommend a complete freeze of the codebase when starting
```
Comment (AntoineVDV) [inline content/developer/howtos/upgrade_custom_db.rst:49]:
```suggestion
every time you change them. This is why we recommend a complete freeze of the codebase when starting
```

File: content/developer/howtos/upgrade_custom_db.rst [hunk 23]
```diff
@@ -0,0 +1,262 @@
...
+the upgrade process. Needless to say, bug fixing is exempt from this recommendation.
+
+Once you have stopped development, it is a good practice to assess the developments made and compare
+them with the features introduced between your current version and the version you are targeting.
+Challenge the developments as much as possible, find functionnal workarounds.
```
Comment (AntoineVDV) [inline content/developer/howtos/upgrade_custom_db.rst:54]:
```suggestion
Challenge the developments as much as possible and find functional workarounds.
```

File: content/developer/howtos/upgrade_custom_db.rst [hunk 24]
```diff
@@ -0,0 +1,262 @@
...
+Removing redundancy between your developments and the standard version of Odoo will lead to an eased
+upgrade process and reduce the technical debt.
```
Comment (AntoineVDV) [inline content/developer/howtos/upgrade_custom_db.rst:56]:
```suggestion
upgrade process and reduce technical debt.
```

File: content/developer/howtos/upgrade_custom_db.rst [hunk 25]
```diff
@@ -0,0 +1,262 @@
...
+
+.. _upgrade_custom/empty_database:
+
+Empty database
+==============
+
+Before working on an upgraded test database, we recommend to make the custom developments work on an
+empty database in the targeted version of your upgrade. 
+This ensures that the customization is compatible with the new version of Odoo, allows to analyse
+how does it behave and interact with the new features, and guarantees that they will not cause any
```
Comment (AntoineVDV) [inline content/developer/howtos/upgrade_custom_db.rst:88]:
```suggestion
how it behaves and interacts with the new features, and guarantees that they will not cause any
```

File: content/developer/howtos/upgrade_custom_db.rst [hunk 26]
```diff
@@ -0,0 +1,262 @@
...
+issue when upgrading the database.
```
Comment (AntoineVDV) [inline content/developer/howtos/upgrade_custom_db.rst:89]:
```suggestion
issues when upgrading the database.
```

File: content/developer/howtos/upgrade_custom_db.rst [hunk 27]
```diff
@@ -0,0 +1,262 @@
...
+
+Making the custom modules work in an empty database also helps avoiding changes and wrong
```
Comment (AntoineVDV) [inline content/developer/howtos/upgrade_custom_db.rst:91]:
```suggestion
Making the custom modules work in an empty database also helps avoid changes and wrong
```

File: content/developer/howtos/upgrade_custom_db.rst [hunk 28]
```diff
@@ -0,0 +1,262 @@
...
+configurations that might be present on the production database (like studio customization,
```
Comment (AntoineVDV) [inline content/developer/howtos/upgrade_custom_db.rst:92]:
```suggestion
configurations that might be present in the production database (like studio customization,
```

File: content/developer/howtos/upgrade_custom_db.rst [hunk 29]
```diff
@@ -0,0 +1,262 @@
...
+customized website pages, email templates or translations). They are not intrinsically related to
+the custom modules and that can raise unwanted issues in this stage of the upgraded process.
+
+To make custom modules work on an empty database we advise to follow these steps:
+
+- :ref:`upgrade_custom/empty_database/modules_installable`
+- :ref:`upgrade_custom/empty_database/test_fixes`
+- :ref:`upgrade_custom/empty_database/clean_code`
+- :ref:`Make standard tests run successfully <upgrade_custom/empty_database/standard_test>`
```
Comment (AntoineVDV) [inline content/developer/howtos/upgrade_custom_db.rst:101]: Numbered list?

File: content/developer/howtos/upgrade_custom_db.rst [hunk 30]
```diff
@@ -0,0 +1,262 @@
...
+--------------
+
+Once there are no more tracebacks when installing the modules, the next step is to test them.
+Even if the custom modules are installable on an empty database, this does not guarantee there are
+no errors during their execution. Because of this, we encourage to test thoroughly all the
+customization to make sure everything is working as expected.
+
+This process will help detect further issues that are not identified during the module installation
+and can only be detected in runtime. For example, deprecated calls to standard python or OWL
+functions, non existing references to standard fields, etc.
```
Comment (AntoineVDV) [inline content/developer/howtos/upgrade_custom_db.rst:134]:
```suggestion
functions, non-existing references to standard fields, etc.
```

File: content/developer/howtos/upgrade_custom_db.rst [hunk 31]
```diff
@@ -0,0 +1,262 @@
...
+
+We recommend to test all the customization, specially the following elements:
```
Comment (AntoineVDV) [inline content/developer/howtos/upgrade_custom_db.rst:136]:
```suggestion
We recommend to test all the customization, especially the following elements:
```

File: content/developer/howtos/upgrade_custom_db.rst [hunk 32]
```diff
@@ -0,0 +1,262 @@
...
+
+
+.. _upgrade_custom/empty_database:
+
+Empty database
+==============
+
+Before working on an upgraded test database, we recommend to make the custom developments work on an
+empty database in the targeted version of your upgrade. 
+This ensures that the customization is compatible with the new version of Odoo, allows to analyse
```
Comment (AntoineVDV) [inline content/developer/howtos/upgrade_custom_db.rst:87]: en-US
```suggestion
This ensures that the customization is compatible with the new version of Odoo, allows to analyze
```

File: content/developer/howtos/upgrade_custom_db.rst [hunk 33]
```diff
@@ -0,0 +1,262 @@
...
+----------------
+
+During the upgrade of the custom modules, you might have to use migration scripts to reflect changes
+from the source code to their corresponding data.
+
+- Any technical data that was renamed during the upgrade of the custom code (models, fields, external
+  identifiers) should be renamed using migration scripts to avoid data loss during the module upgrade.
+- Data from standard models removed from the source code of the newer Odoo version and from the
+  database during the standard upgrade process might need to be recovered from the old model table
+  if it is stil present.
```
Comment (AntoineVDV) [inline content/developer/howtos/upgrade_custom_db.rst:206]:
```suggestion
  if it is still present.
```

File: content/developer/howtos/upgrade_custom_db.rst [hunk 34]
```diff
@@ -0,0 +1,262 @@
...
+
+Migration scripts can also be used to:
+
+- Ease the processing time of an upgrade. For example, to store the value of computed stored fields
+  on models with an excesive amount of records by using SQL queries.
```
Comment (AntoineVDV) [inline content/developer/howtos/upgrade_custom_db.rst:211]:
```suggestion
  on models with an excessive number of records by using SQL queries.
```

File: content/developer/howtos/upgrade_custom_db.rst [hunk 35]
```diff
@@ -0,0 +1,262 @@
...
+- Correct faulty data or wrong configurations.
+
+.. _upgrade_custom/upgraded_database/test_custom:
+
+Test the custom modules
+-----------------------
+
+To make sure the custom modules work properly with your data in the upgraded database, they need to
+be tested as well. This helps ensure both the standard and the custom data stored in the database
+are consistent and nothing was lost during the upgrade process.
```
Comment (AntoineVDV) [inline content/developer/howtos/upgrade_custom_db.rst:223]:
```suggestion
are consistent and that nothing was lost during the upgrade process.
```

File: content/developer/howtos/upgrade_custom_db.rst [hunk 36]
```diff
@@ -0,0 +1,262 @@
...
+
+Things to pay attention to:
+
+- Views not working: During the upgrade, if a view causes issues because of it's content, it gets
```
Comment (AntoineVDV) [inline content/developer/howtos/upgrade_custom_db.rst:227]:
```suggestion
- Views not working: During the upgrade, if a view causes issues because of its content, it gets
```

File: content/developer/howtos/upgrade_custom_db.rst [hunk 37]
```diff
@@ -0,0 +1,262 @@
...
+  disabled. You can find the information of disabled views on the :ref:`Upgraded report
```
Comment (AntoineVDV) [inline content/developer/howtos/upgrade_custom_db.rst:228]:
```suggestion
  disabled. You can find the information on disabled views on the :ref:`Upgraded report
```

File: content/developer/howtos/upgrade_custom_db.rst [hunk 38]
```diff
@@ -0,0 +1,262 @@
...
+  <upgrade/upgrade_report>`. This views need to be activated again. To achieve this, we recommend
```
Comment (AntoineVDV) [inline content/developer/howtos/upgrade_custom_db.rst:229]:
```suggestion
  <upgrade/upgrade_report>`. This view needs to be activated again. To achieve this, we recommend
```

File: content/developer/howtos/upgrade_custom_db.rst [hunk 39]
```diff
@@ -0,0 +1,262 @@
...
+round of testing to assess the database usability and detect any issue that might have gone
+unnoticed in previous tests. For further information about testing the upgraded database, check 
+:ref:`upgrade/test_your_db`.
+
+As mentioned in :ref:`upgrade/upgrade-prod`, both standard upgrade scripts and your database are
+constantly evolving. Therefore it is highly recommended to frequently request new upgraded test
+databases and ensure that the upgrade process is still successful.
+
+In adition to that, make a full rehearsal of the upgrade process the day before upgrading the
+production database to avoid an undesired behavior during the upgrade and to detect any issue that
```
Comment (AntoineVDV) [inline content/developer/howtos/upgrade_custom_db.rst:252]:
```suggestion
production database to avoid undesired behavior during the upgrade and to detect any issue that
```

File: content/developer/howtos/upgrade_custom_db.rst [hunk 40]
```diff
@@ -0,0 +1,262 @@
...
+might have ocurred with the migrated data.
```
Comment (AntoineVDV) [inline content/developer/howtos/upgrade_custom_db.rst:253]:
```suggestion
might have occurred with the migrated data.
```

File: content/developer/reference/cli.rst [hunk 41]
```diff
@@ -669,18 +669,41 @@ Here is a sample file:
...
 
    $ odoo-bin shell
 
 .. option:: --shell-interface (ipython|ptpython|bpython|python)
 
-    Specify a preferred REPL to use in shell mode.
+   Specify a preferred REPL to use in shell mode. This shell is started with the `env` variable
+   already initialized to be able to access the ORM and other Odoo modules.
+
+.. example:: Example of shell usage
```
Comment (AntoineVDV) [inline content/developer/reference/cli.rst:684]:
```suggestion
.. example::
```

Comment (AntoineVDV) [inline content/developer/reference/cli.rst:684]: I'd move the example before the `shell-interface` option; readers want to know how it works before diving into the details.

File: content/developer/reference/cli.rst [hunk 42]
```diff
@@ -669,18 +669,41 @@ Here is a sample file:
...
+         ...:
+
+      In [4]: env.cr.commit()
+
+   .. important::
+      By default, the shell is running in transaction mode. This means that any change made to the
+      database is rolled back when exiting the shell. To commit changes, use `env.cr.commit()`.
+
+.. seealso::
+   :ref:`Environment documentation <reference/orm/environment>`
```
Comment (AntoineVDV) [inline content/developer/reference/cli.rst:706]:
```suggestion
   :ref:`reference/orm/environment`
```

**Review by AntoineVDV — CHANGES_REQUESTED** [review]
Hello, interesting doc :slightly_smiling_face:

I put my review comments for the /developer part of the diff below. None of them are really blocking, so feel free to apply or discard.

Also,
- The [previous rewrite of the upgrade documentation](https://github.com/odoo/documentation/pull/6274) targeted 16.0. Wouldn't it make sense to target the same branch with this PR so that partners browsing the documentation in 16.0 can find the How-to guide too?
- I added a codeowner rule for the new file so the `ci/codeowner_coverage` should be green, but you must also fix whatever is making the `ci/documentation_guidelines` red.

File: content/developer/howtos/upgrade_custom_db.rst (hunk 20 above)
Comment (augustoPerez) [reply content/developer/howtos/upgrade_custom_db.rst:39]: Would you recommend to do something similar to the [Bugfix update page]?(https://www.odoo.com/documentation/17.0/administration/maintain/update.html)
image

File: content/developer/howtos/upgrade_custom_db.rst [hunk 43]
```diff
@@ -0,0 +1,262 @@
...
+.. _upgrade_custom/upgrade_custom_db:
+
+===============================
+Upgrading a customized database
+===============================
+
+Upgrading to a new version of Odoo can be challenging, especially if the database you work on
+contains custom modules. This page intent is to explain the technical process of upgrading a
+database with customized modules. For a standard overview, please refer to the
+:doc:`Upgrade documentation </administration/upgrade>`.
```
[the remark this answers is not in the document]
Comment (augustoPerez) [reply content/developer/howtos/upgrade_custom_db.rst]: I don't see the need for this, it's not detailed in any other documentation page and no Odoo developer reads all the documentation anyway (you go on it only if you need it for a specific reason at a certain point).

File: content/developer/howtos/upgrade_custom_db.rst (hunk 20 above)
Comment (AntoineVDV) [reply content/developer/howtos/upgrade_custom_db.rst:34]: Yeah, it's an idea. The thing is that very few people actually read documentation through and through; they scroll until they find something that looks like the solution to their problem. Introducing steps will help clarify that sections must be read in a sequence.

**Comment by AntoineVDV** [comment]
In general, version-independent documentation PRs should target the lowest supported branch (15.0 here). Since the previous PR targeted 16.0, I recommend targeting 16.0 too, and in less than one year, 15.0 will be deprecated anyway. The reason is that users accessing the documentation from their Odoo database will be redirected to the version of the documentation that matches the version of their database.

**Comment by augustoPerez** [comment]
@AntoineVDV is it okay not to rename the brach when targeting 16.0?

File: content/developer/howtos/upgrade_custom_db.rst (hunk 20 above)
Comment (augustoPerez) [reply content/developer/howtos/upgrade_custom_db.rst]: Sound good to me, I didn't think of that
Thanks for pointing it out! :+1:

**Comment by AntoineVDV** [comment]
> @AntoineVDV is it okay not to rename the brach when targeting 16.0?

Yeah it's fine; it's only helpful for runbot until the PR is created, then the PR's target branch is considered the base branch.

**Review by AntoineVDV — APPROVED** [review]
Just squash the commits and we're good to go.

**Comment by augustoPerez** [comment]
@AntoineVDV commits squashed :wink:

**Comment by AntoineVDV** [comment]
@robodoo r+
